### ISSUE TITLE
Siemens DICOM type

### ISSUE THREAD

**Issue by barna21** [fo-dicom/fo-dicom#1339]
Hello,
These days I have discovered a new type of DICOM that cannot be processed by Fo-dicom.
The application that can open it is from Siemens and I didn't even manage to convert it to ExplicitVRLittleEndian.
Can you suggest something to me?

Thank you in advance!
![syngo fastView](https://user-images.githubusercontent.com/33651659/155710920-c38b633b-a0b6-4998-b111-da2b8ecf056f.png)

**Comment by mrbean-bremen** [comment]
Can you show the DICOM header, at least the Transfer Syntax UID and the SOP Class UID?

**Comment by barna21** [comment]
I uploaded some files for analysis.
[https://drive.google.com/drive/folders/1-1v5T0ACnOInm7EZmwKtvwgfXz-DmwtS?usp=sharing](https://drive.google.com/drive/folders/1-1v5T0ACnOInm7EZmwKtvwgfXz-DmwtS?usp=sharing)

**Comment by mrbean-bremen** [comment]
First: the data seems not be anonymized, it may not be a good idea to provide a public link to it.
Second: these are EnhancedSRStorage data, saved as LittleEndianExplicit, I think fo-dicom should be able to read the data. Can you show the code you are using?

**Comment by barna21** [comment]
I deleted the link in the meantime. I couldn't anonymize as long as I couldn't process the data. I tried to open them with radiant and it doesn't open there either ...

**Comment by mrbean-bremen** [comment]
I'm not sure what you mean by "open". You are writing "cannot be processed by fo-dicom" - can you elaborate on what you did and what failed?

**Comment by mrbean-bremen** [comment]
Ok, I take that back about readable files: only the first file is readable (the SR, as I wrote), the others (MR images, by the look of it), have some invalid DICOM that probably cannot be processed by most frameworks (assuming that they have been downloaded correctly). This is the output from dcmdump:
```
W: DcmSequenceOfItems: Reached end of stream before the end of sequence PixelData (7fe0,0010)
E: dcmdump: Sequence Delimitation Item missing: reading file: 57152647
```
I will try to check it later today.

**Comment by mrbean-bremen** [comment]
It looks like the pixel data itself is invalid. The transfer syntax is JPEG Lossless, Non-Hierarchical, First-Order Prediction (Process 14), and the DICOM header looks mostly ok, as far as I can see. As the pixel data is not read by fo-dicom itself, but by fo-dicom.Codecs, I'm not sure there is anything we can do here.
Anyway this is clearly invalid data, and given that no reader I tried can read it, it is not something that can be easily worked around - I think you are out of luck here (not the first time I see invalid DICOM written by this specific vendor that can only be read by software of the same vendor, sigh...).

For reference, some of the the relevant tags in the header:
```
(0002, 0002) Media Storage SOP Class UID         MR Image Storage
(0002, 0010) Transfer Syntax UID                 JPEG Lossless, Non-Hierarchical, First-Order Prediction (Process 14 [Selection Value 1])
(0002, 0013) Implementation Version Name         SIEMENS
(0008, 0005) Specific Character Set              ISO_IR 100
(0008, 0008) Image Type                          DERIVED\PRIMARY\M\NORM\WATER\DIS2D\MFSPLIT
(0008, 0016) SOP Class UID                       MR Image Storage
(0008, 0060) Modality                            MR
(0008, 0070) Manufacturer                        Siemens
(0008, 1090) Manufacturer's Model Name           MAGNETOM Altea
(0008, 2111) Derivation Description              Compress Pegasus JPEG Lossless
(0018, 1020) Software Versions                   syngo MR XA20
(0028, 0002) Samples per Pixel                   1
(0028, 0004) Photometric Interpretation          MONOCHROME2
(0028, 0010) Rows                                280
(0028, 0011) Columns                             320
(0028, 0100) Bits Allocated                      16
(0028, 0101) Bits Stored                         12
(0028, 0102) High Bit                            11
(0028, 0103) Pixel Representation                0
(0028, 2110) Lossy Image Compression             00
```

**Comment by scotthartzell** [comment]
I have the same issue, not faulty data, something like a missing 'stop' bit in the PixelData itself. Siemens Syngo MRI

**Comment by Muttleskovic** [comment]
Any workaround on this thread? Both RadiAnt and MicroDicom have no issue on displaying this type of dicom images.
Sadly can't upload a anonymized image because on export/anon the file it's also fixed and can be open/render with fo-dicom.

**Comment by amoerie** [comment]
If someone is willing to anonymize a DICOM file (perhaps using a hex editor like HxD - be sure to replace names with strings of the same length) then we might be able to debug this and perhaps provide a workaround. 
Until then, I hope you understand there is not much that we as fo-dicom contributors can do.

**Comment by Muttleskovic** [comment]
I've managed to anonymize the image using Alexander suggested method with Hex Editor
https://file.io/63Bq4Chprs2t
Edit: Not sure what happen, maybe the link got deleted after I've tested if can be downloaded
New link : https://we.tl/t-WRyNXI1uAz

**Comment by amoerie** [comment]
Ah, I deleted my comment because I realized @mrbean-bremen was already in possession of the faulty files so maybe he could forward them privately to me, but this is even better. :-) 

Edit: looks like the link does not work @Muttleskovic ?

**Comment by Muttleskovic** [comment]
My guess is that the pixel data is inside a custom tag (0021,1019) since it has the higher length, but not sure how to access it or use it as pixel data
![image](https://github.com/fo-dicom/fo-dicom/assets/20122218/b8bf0ee4-4859-4a39-bd4d-2583aac0e6b3)

**Comment by amoerie** [comment]
The DICOM file says that the length of the Pixel Data tag will be uint.MaxValue. This means "undefined length" (TIL) and it means we will parse the ensuing bytes as a fragment sequence.

![image](https://github.com/fo-dicom/fo-dicom/assets/2063584/a7bc5f23-1c61-493f-92df-f3529bf8069b)

Since the transfer syntax is JPEG Lossless, an explicit VR syntax, it first says that the Pixel Data has the VR type 'OB' (Other Byte), then 2 bytes are skipped and then the next 4 bytes are read into an unsigned integer.

![image](https://github.com/fo-dicom/fo-dicom/assets/2063584/a5ec16f7-f7c8-4b08-9428-fc48d816138c)

We do seem to support undefined length, we try to parse the pixel data as fragments. But that fails somehow. ~~Still investigating...~~

![image](https://github.com/fo-dicom/fo-dicom/assets/2063584/d3eaee07-9ab0-4419-a1d5-a9f951393b2b)

So, I debugged a little further, and inside the fragment sequence parsing, it first retrieves the tag (offset 86236) group, which is FFFE and the tag element which is E000, or simply put DicomTag.PixelData. (Remember that the bytes are swapped, so the original file contains FE FF 00 E0)

Then, the length of the fragment sequence is read from the next 4 bytes into a uint. As you can see, the next 4 bytes are all zero, so the length of the fragment sequence will be 0. 

The first fragment like this is empty (and this is allowed), but the second fragment contains all of the pixel data. It looks like fo-dicom simply fails to call `_observer.OnEndFragmentSequence();` if there is no SequenceDelimitationItem at the end...

**Comment by amoerie** [comment]
@barna21 @scotthartzell @Muttleskovic I think I have found a satisfying solution to parsing these DICOM files, please see #1718

**Comment by Muttleskovic** [comment]
Thank you very much for your support, not sure how I can update local libraries, but i guess the fix will be available in next nuget source version, 5.1.2 ATM

**Comment by amoerie** [comment]
> Thank you very much for your support, not sure how I can update local libraries, but i guess the fix will be available in next nuget source version, 5.1.2 ATM

That is correct. Your options are 

A) Wait for the next fo-dicom release (this usually takes a while, sorry)
B) Compiling the source code yourself (you could use a git submodule for example)
